text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+        * Alternatively `where <T as Trait>::some_fn(..): ~const` or `where <T as Trait>::some_fn \ {const}`.
+        * Very verbose (need to specify arguments and return type).
+        * Want short hand sugar anyway to make it trivial to change a normal function to a const function by just adding some minor annotations.
+        * Significantly would delay const trait stabilization (by years).
+        * Usually requires editing the trait anyway, so there's no "can constify impls without trait author opt in" silver bullet.
+    * New RTN-like per-method bounds: `T: Trait<some_fn(_): ~const>`.
+        * Unclear if soundly possible.
+        * Unclear if possible without incurring significant performance issues for all code (may need tracking new information for all functions out there).
+        * Still requires editing traits.
+        * Still want the `~const Trait` sugar anyway.
+
+## Should we start out by allowing only const trait declarations and const trait impls
+
+We do not need to immediately allow using methods on generic parameters of const fn, as a lot of const code is nongeneric.
+
+The following example could be made to work with just const traits and const trait impls.
+
+```rust
+const fn foo() {
+    let a = [1, 2, 3];
+    let b = [1, 2, 4];
+    if a == b {}
+}
+```
+
+Things like `Option::map` could not be made const without const trait bounds, as they need to actually call the generic `FnOnce` argument.
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Migrate to `~const fn`
+
+`const fn` and `const` items have slightly different meanings for `const`:
+
+`const fn` can also be called at runtime just fine, while the others are always const
+contexts and need to be evaluated by the const evaluator.
+
+Additionally `const Trait` bounds have a third meaning (the same as `const Trait` in `impl const Trait for Type`):
+
+They can be invoked at compile time, but also in `const fn`.
+
+While all these meanings are subtly different, making their differences more obvious will not make them easier to understand.
+All that changing to `~const fn` would achieve is that folk will add the sigil when told by the compiler, and complain about
+having to type a sigil, when there is no meaning for `const fn` without a sigil.
+
+While I see the allure from a language nerd perspective to give every meaning its own syntax, I believe it is much more practical to
+just call all of these `const` and only separate the `~const Trait` bounds from `const Trait` bounds.
+
+## `const fn()` pointers
+
+Just like `const fn foo(x: impl ~const Trait) { x.method() }` and `const fn foo(x: &dyn ~const Trait) { x.method() }` we want to allow
+`const fn foo(f: const fn()) { f() }`.
+
+There is nothing design-wise blocking function pointers and calling them, they mainly require implementation work and extending the
+compiler's internal type system representation of a function signature to include constness.
```

Comment:
Are we sure about that? I’m worried about subtyping issues. One nice property upheld by this RFC is “you should never run into `const`-related issues if you are not actually executing code in `const`”, but I’m not sure how easy that will be to preserve with function pointers or `dyn`.

For example, if I have an `fn foo(x: impl Trait<Assoc = fn()>)`, can I pass in an `impl Trait<Assoc = const fn()>`? There are situations where that might be necessary, but also situations where it can’t be allowed. Also, what is the behavior of `TypeId`? Etc.

Reply:
`const fn()` and `fn()` will need to be different have `TypeId`s, at least if we want to support `dyn const Any` in the future.

Reply:
`const fn()` and `fn()` will need to be different have `TypeId`s, at least if we want to support `dyn const Any` in the future.